### ISSUE TITLE
Another template for VIRTUAL_PATHs instead of VIRTUAL_HOSTs

### ISSUE THREAD

**Issue by andrefernandes** [nginx-proxy/nginx-proxy#33]
What about offering other templates for different needs?

I am using the template below to generate several "location" entries for each server. This results on a single domain with a context root for each container (more like port redirection). Serves me well for my single-server Digital Ocean droplet. 

You can have several templates available in this project, it is just a matter of mounting the chosen template into "/app/nginx.tmpl" when running the proxy.

---

```
server {
    listen 80 default_server;
    server_name _; # This is just an invalid value which will never trigger on a real hostname.
    error_log /proc/self/fd/2;
    access_log /proc/self/fd/1;
    return 503;
}

{{ range $host, $containers := groupByMulti $ "Env.VIRTUAL_HOST" "," }}

server {
    gzip_types text/plain text/css application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;

    server_name {{ $host }};
    proxy_buffering off;
    error_log /proc/self/fd/2;
    access_log /proc/self/fd/1;

{{ range $index, $value := $containers }}
{{ $location := $value.Env.VIRTUAL_PATH }}
{{ $address := index $value.Addresses 0 }}
# {{$location}}

    location /{{$location}} {
            proxy_pass http://{{ $address.IP }}:{{ $address.Port }};
            include /etc/nginx/proxy_params;

            # HTTP 1.1 support
            proxy_http_version 1.1;
            proxy_set_header Connection "";
    }
{{ end }}

}

{{ end }}
```

**Comment by trinitronx** [comment]
For anyone running this in front of a [docker registry](https://registry.hub.docker.com/_/registry/) container...

Now that the [docker registry v2.0](http://blog.docker.com/2015/04/faster-and-better-image-distribution-with-registry-2-0-and-engine-1-6/) is out, and they've pretty much done away with `/v1/` API backwards compatibility, this feature / use case is definitely needed ;-)

They're [currently recommending to run a `v1` and `v2`](https://docs.docker.com/registry/deploying/#configure-nginx-with-a-v1-and-v2-registry) registry [in parallel](https://github.com/docker/distribution/blob/62b70f951f30a711a8a81df1865d0afeeaaa0169/docs/deploying.md#configure-nginx-with-a-v1-and-v2-registry) with an nginx proxy in front.  Their [nginx config](https://github.com/docker/distribution/blob/62b70f951f30a711a8a81df1865d0afeeaaa0169/contrib/compose/nginx/registry.conf#L22) just passes the requests for `/v1/` and `/v2/` to the different backend containers.

**Comment by md5** [comment]
Getting this working for the `/v1`/`/v2` case is pretty easy and already works in [https://github.com/jwilder/nginx-proxy/compare/master...appropriate:multiple-paths](my `multiple-paths` branch).  Getting it working in such a way that it's mergeable in the general case is quite a bit harder. The underlying `docker-gen` utility would have to include a `sort*` command that could do a reverse sort by longest path segments. There would also have to be a provision for providing a sane (and secure) response for `/` if no backing container provided a `VIRTUAL_PATH=/`.

**Comment by PavelVanecek** [comment]
Because of reasons, github points md5's link above to incorrect address. The actual link is: https://github.com/jwilder/nginx-proxy/compare/master...appropriate:multiple-paths

**Comment by md5** [comment]
I just took a look at the Nginx docs and found that `location` matches don't actually happen in the order they're defined. Nginx already does a "longest prefix" algorithm:

> To find location matching a given request, nginx first checks locations defined using the prefix strings (prefix locations). Among them, the location with the longest matching prefix is selected and remembered. Then regular expressions are checked, in the order of their appearance in the configuration file. The search of regular expressions terminates on the first match, and the corresponding configuration is used. If no match with a regular expression is found then the configuration of the prefix location remembered earlier is used.

So I think the only thing that actually needs to be fixed to make my `multiple-paths` branch usable is to make sure that a 503 error is returned for `/` when no `location` matches it.

**Comment by lonix** [comment]
I Would love this feature.

**Comment by md5** [comment]
I just took a look at getting my https://github.com/jwilder/nginx-proxy/compare/master...appropriate:multiple-paths branch back in shape and it's turning out to be quite painful...

There have been quite a few changes in `nginx.tmpl` since I last updated the branch and the logic in the template is getting more and more complicated. It will only get more complicated if I add this or if some of the open feature requests get implemented (e.g. making the HTTP to HTTPS redirect conditional).

Trying to do a direct rebase seems to be more trouble than it's worth. I think the only viable approach is to reproduce the `VIRTUAL_PATH` changes from scratch.

**Comment by md5** [comment]
I guess I spoke too soon. I managed to get the branch closer to a mergeable state, though it still doesn't handle the 503 issue I mentioned in my comment from July: https://github.com/jwilder/nginx-proxy/issues/33#issuecomment-125935125

Before doing any more work on that branch, I'd like to see #244 merged. I think it still needs to have the additional lines fixed that are mentioned in this comment: https://github.com/jwilder/nginx-proxy/pull/244#issuecomment-145524471

I'd also like to see the whitespace in the file made consistent again, but that's not a huge deal.

**Comment by md5** [comment]
I finally opened what I think is a mergeable PR for this feature over at https://github.com/jwilder/nginx-proxy/pull/254

Please discuss the patch I've offered at that PR if you're interested in this feature.

**Comment by andrefernandes** [comment]
Sure!

De: Mike Dillon [mailto:notifications@github.com]
Enviada em: sexta-feira, 9 de outubro de 2015 00:50
Para: jwilder/nginx-proxy <nginx-proxy@noreply.github.com>
Cc: Andre de Oliveira Fernandes <andre.oliveira@bcb.gov.br>
Assunto: Re: [nginx-proxy] Another template for VIRTUAL_PATHs instead of VIRTUAL_HOSTs (#33)


I finally opened what I think is a mergeable PR for this feature over at #254<https://github.com/jwilder/nginx-proxy/pull/254>

Please discuss the patch I've offered at that PR if you're interested in this feature.

**Comment by buchdag** [comment]
This features has been added to the `dev` branch with #1607 

It's available now on the `dev` images `nginxproxy/nginx-proxy:dev` and `nginxproxy/nginx-proxy:dev-alpine`. Testing and feedback are welcome, the docs for the feature is [here](https://github.com/nginx-proxy/nginx-proxy/tree/dev#path-based-routing) and there is [another PR](https://github.com/nginx-proxy/nginx-proxy/pull/1699) opened for additional improvements and comments.
